File: baekjoon/data_structure/7785_r.py
import sys
from typing import Dict
input = sys.stdin.readline


def solution(access_log: Dict[str, int]):
    cur_entered_workers = [k for k, v in access_log.items() if v > 0]

    for w in sorted(cur_entered_workers, reverse=True):
        print(w)


if __name__ == '__main__':
    input_num = int(input())
    access_log = {}

    for _ in range(input_num):
        name, enter_or_leave = input().split()
        if enter_or_leave[0] == 'l':
            access_log[name] = 0
        else:
            access_log[name] = 1

    solution(access_log)

File: baekjoon/data_structure/7785_r.js
// const solution = input => {
//   const workerMap = new Map();

//   for (const log of input.slice(1)) {
//     const [name, desc] = log.split(' ');

//     desc.startsWith('e') ? workerMap.set(name, 1) : workerMap.set(name, 0);
//   }

//   const curWorker = [];
//   for (const [key, value] of workerMap) if (value) curWorker.push(key);

//   return curWorker.sort().reverse().join('\n');
// };

const solution = input => {
  const curWorker = new Set();

  for (const log of input.slice(1)) {
    const [name, desc] = log.split(' ');

    desc.startsWith('e') ? curWorker.add(name) : curWorker.delete(name);
  }

  return Array.from(curWorker).sort().reverse().join('\n');
};

function main() {
  const fs = require('fs');
  const input = fs.readFileSync('./baekjoon/data_structure/7785.txt').toString().trim().split('\n');

  console.log(solution(input));
}

if (require.main === module) main();
